src/main/kotlin/org/rust/ide/intentions/ImportNameIntention.kt

```diff
@@ -0,0 +1,232 @@
+        // we need to do it because we can use direct child items of any super mod with any visibility
+        val lca = ourSuperMods.find { it in superMods }
+
+        val shouldBePublicMods = if (lca == null) {
+            if (!isPublic) return false
+            ourSuperMods
+        } else {
+            // if current item is direct child of some ancestor of `mod` then it can be not public
+            if (parentMod == lca) return true
+            if (!isPublic) return false
+            ourSuperMods.takeWhile { it != lca }.dropLast(1)
+        }
+        return shouldBePublicMods.all { it.isPublic }
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val (path, candidates) = ctx
+
+        if (candidates.size == 1) {
+            importItem(project, candidates.first(), path)
+        } else {
+            chooseItemAndImport(project, editor, candidates, path)
+        }
+    }
+
+    private fun chooseItemAndImport(project: Project, editor: Editor, items: List<RsQualifiedNamedElement>, originalPath: RsPath) {
+        // TODO: sort items in popup
+        val step = object : BaseListPopupStep<RsQualifiedNamedElement>("Item to Import", items) {
+            override fun isAutoSelectionEnabled(): Boolean = false
+            override fun isSpeedSearchEnabled(): Boolean = true
+            override fun hasSubstep(selectedValue: RsQualifiedNamedElement?): Boolean = false
+
+            override fun onChosen(selectedValue: RsQualifiedNamedElement?, finalChoice: Boolean): PopupStep<*>? {
+                if (selectedValue == null) return PopupStep.FINAL_CHOICE
+                return doFinalStep { importItem(project, selectedValue, originalPath) }
+            }
+
+            override fun getTextFor(value: RsQualifiedNamedElement): String =
+                ObjectUtils.assertNotNull(value.qualifiedName)
+
+            override fun getIconFor(value: RsQualifiedNamedElement): Icon? = value.getIcon(0)
+        }
+        val popup = object : ListPopupImpl(step) {
+            override fun getListElementRenderer(): ListCellRenderer<*> {
+                val baseRenderer = super.getListElementRenderer() as PopupListElementRenderer<Any>
+                val psiRenderer = RsElementCellRenderer()
+                return ListCellRenderer<Any> { list, value, index, isSelected, cellHasFocus ->
+                    val panel = JPanel(BorderLayout())
+                    baseRenderer.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus)
+                    panel.add(baseRenderer.nextStepLabel, BorderLayout.EAST)
+                    panel.add(psiRenderer.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus))
+                    panel
+                }
+            }
+        }
+        NavigationUtil.hidePopupIfDumbModeStarts(popup, project)
+        popup.showInBestPositionFor(editor)
+    }
```

Comment:
I expect auto import machinery to grow significantly over time, so we might want to create an `import` package and move all UI to a separate file.

Reply:
See also src/main/kotlin/org/rust/lang/refactoring/introduceVariable/ui.kt for an example of mock-based testing infrastructure for similar UI-dialogs.